File: 프로그래머스/unrated/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> answer(2);
    answer[0] = -1;
    answer[1] = -1;
    int start = 0, end = 0;
    int s = sequence[start];
    while (start <= end && end < sequence.size()) {
        if (s == k) {
            if (answer[0] == -1 || answer[1] - answer[0] > end - start) {
                answer[0] = start;
                answer[1] = end;
            }
            end++;
            s += sequence[end];
        }
        else if (s < k) {
            end++;
            s += sequence[end];
        }
        else {
            s -= sequence[start];
            start++;
        }
    }
    return answer;
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.java
class Solution {
    public int[] solution(int[] sequence, int k) {
        int[] res = new int[]{0, sequence.length - 1};
        int left = 0, right = 0, totalSum = sequence[0];
        while (right < sequence.length) {
            if (totalSum == k) {
                if (res[1] - res[0] > right - left)
                    res = new int[]{left, right};
                totalSum -= sequence[left++];
            }
            else if (totalSum < k) {
                right++;
                if (right < sequence.length)
                    totalSum += sequence[right];
            }
            else {
                totalSum -= sequence[left++];
            }
        }
        return res;
    }
}